src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/ProvenanceHeaderBehavior.cs

```diff
@@ -0,0 +1,112 @@
+    {
+        private readonly FhirJsonParser _fhirJsonParser;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly IMediator _mediator;
+        private IProvenanceHeaderState _state;
+
+        public ProvenanceHeaderBehavior(FhirJsonParser fhirJsonParser, IHttpContextAccessor httpContextAccessor, IMediator mediator, IProvenanceHeaderState state)
+        {
+            EnsureArg.IsNotNull(fhirJsonParser, nameof(fhirJsonParser));
+            EnsureArg.IsNotNull(httpContextAccessor, nameof(httpContextAccessor));
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+
+            _fhirJsonParser = fhirJsonParser;
+            _httpContextAccessor = httpContextAccessor;
+            _mediator = mediator;
+            _state = state;
+        }
+
+        public async Task<UpsertResourceResponse> Handle(ConditionalUpsertResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<UpsertResourceResponse> next)
+            => await GenericHandle(next, cancellationToken);
+
+        public async Task<UpsertResourceResponse> Handle(ConditionalCreateResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<UpsertResourceResponse> next)
+            => await GenericHandle(next, cancellationToken);
+
+        public async Task<UpsertResourceResponse> Handle(UpsertResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<UpsertResourceResponse> next)
+            => await GenericHandle(next, cancellationToken);
+
+        public async Task<UpsertResourceResponse> Handle(CreateResourceRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<UpsertResourceResponse> next)
+            => await GenericHandle(next, cancellationToken);
+
+        private async Task<UpsertResourceResponse> GenericHandle(RequestHandlerDelegate<UpsertResourceResponse> next, CancellationToken cancellationToken)
+        {
+            if (_state.Intercepted)
+            {
+                return await next();
+            }
+
+            _state.Intercepted = true;
+            Provenance provenance = GetProvenanceFromHeader();
+            var response = await next();
+
+            if (response != null && provenance != null)
+            {
+                // Set target to provided resource.
+                provenance.Target = new System.Collections.Generic.List<ResourceReference>()
+                {
+                    new ResourceReference($"{response.Outcome.RawResourceElement.InstanceType}/{response.Outcome.RawResourceElement.Id}"),
+                };
+
+                // Create Provenance resource.
+                // TODO: It should probaby go through controller to trigger audit events, but it's quite tricky to do now.
+                await _mediator.Send<UpsertResourceResponse>(new CreateResourceRequest(provenance.ToResourceElement()), cancellationToken);
```

Comment:
Is it possible that this could fail, perhaps with a more in depth validation beyond what the FhirJsonParser performs?  If it did fail, should we catch and throw a ProvenanceHeaderException here as well?

Reply:
I don't think so.
So there is lot of things can happen here, persistence storage is down, validation failure, maybe some weird misconfiguration with search parameters, who knows... And all this cases will throw thier own exception which I need to catch and wrap in ProvenanceHeaderException? I would rather just let them go to the top, I'm not sure it's there is much benefit I would gain from throwing specific exception.